add event manage api to Bigpipe Object

naiteluo opened the pull request:
Add Event manage api to Bigpipe object.

Dispatch events: `beforepageletload`, `pageletstyleloaded`,`pageletscriptloaded`, so that we can modify pagelet or do things in different times of the pagelet's life circle.

naiteluo committed:
add event config to bind event to pagelet

naiteluo committed:
fix #2 implicitly declared issue

naiteluo committed:
Bigpipe event add/remote/dispatch; Test cases modified.

naiteluo committed:
add event before pagelet load

leeyeh committed:
Stop broadcast 'pageletload' event, which should have been handled in Bigpipe 'pagelet' event callback.

naiteluo committed:
modify beforepageletload event's order

leeyeh commented:
我来 rebase 一下。

cloverzero commented on lib/bigpipe.js:
```diff
@@ -617,12 +623,64 @@
+     */
+    Bigpipe.removeEventListener = function (name, handler) {
+        if (_BigpipeEventHandlers[name]) {
+            delete _BigpipeEventHandlers[name];
```
这不是把所有name相同的handler全删掉了嘛

cloverzero commented on lib/bigpipe.js:
```diff
@@ -617,12 +623,64 @@
+        var callbacks = _BigpipeEventHandlers[name];
+        if (callbacks) {
+            for (var i = callbacks.length - 1; i >= 0; i--) {
+                callbacks[i].call(Bigpipe, data);
```
为什么要倒序调用？

cloverzero commented on lib/bigpipe.js:
```diff
@@ -643,7 +701,8 @@
         Promise.all(loadPagelets).then(function () {
-            Bigpipe.broadcast('pageload');
+            // tell Bigpipe that the page is loaded
+            Bigpipe.dispatchEvent('pageload', pagelets);
```
我觉得pageloaded更合适一些

leeyeh committed:
Fix code review issues.

- optimize some Promise.
- Bigpipe and pagelet event callbacks are executed in correct order.
